WIP: mutually-compatible features checkpointing

jngrad opened the pull request:
Fixes #2436, Fixes #2365, Fixes #333 

Description of changes:
 - use a single checkpoint script to test multiple mutually-compatible features
 - define semicolon-delimited lists of feature combinations in `/testsuite/python/CMakeLists.txt`
    - each combination is a hyphen-delimited list of mutually-compatible features
    - use a dot to add flags, e.g. `lb.gpu-p3m.cpu-lj` tests for LB on the GPU together with P3M on the CPU and LJ
    - for non-binary checkpointing, `lb.gpu-lb.ascii-p3m.cpu-lj` could be used, or `lb.gpu.ascii-p3m.cpu-lj` after a rewrite of the `modes` parser

@jainka, @KaiSzuttor 

Edit: this PR also fixes #1848, fixes #2555, fixes #2564

jngrad committed:
Enable template test scripts

jngrad committed:
Add flags to features

jngrad committed:
Simplify template test scripts

Commit:
Updated the checkpoint tests

Using ASCII Vs Binary checkpoint

jngrad committed:
Formatting

KaiSzuttor commented:
What is the status here?

jngrad committed:
Remove comments

jngrad commented:
The checkpointing tests now cover LB CPU+GPU (#333), ASCII+binary checkpointing (#2365) and are more flexible (#2436). Provided CI passes, I'd say we're good to merge and close all 3 tickets.

KaiSzuttor commented on testsuite/python/CMakeLists.txt:
```diff
@@ -59,8 +63,18 @@ function(PYTHON_TEST)
 
-python_test(FILE save_checkpoint.py MAX_NUM_PROC 4)
-python_test(FILE test_checkpoint.py MAX_NUM_PROC 4 DEPENDS save_checkpoint)
+foreach(TEST_COMBINATION lb.cpu-p3m.cpu-lj;lb.gpu-p3m.cpu-lj)
```
Could you please add a comment how to add new tests?

jngrad replied:
Done.

jngrad committed:
Add short explanation

jngrad commented:
Ticket #1848 can also be closed after this merge.

fweik commented:
@KaiSzuttor @jngrad What is still missing here?

KaiSzuttor committed:
Merge branch 'python' of https://github.com/espressomd/espresso into fix-2436

jngrad committed:
Merge branch 'python' into fix-2436

jngrad committed:
Fix test: feature, precision, DeprecationWarning

jngrad committed:
Repair LB interface

jngrad commented:
@fweik @KaiSzuttor We should be able to merge cleanly now.

jngrad committed:
Improve exception handling in LB checkpointing

- don't read from a file after fclose(): undefined behavior
- throw on first error with any read operation
- throw if checkpoint file contains more data than expected
- throw if lattice dimensions don't match
- test exception handling

jngrad committed:
Test a non-cubic box

jngrad committed:
Formatting

jngrad commented:
@KaiSzuttor in the current `python` branch, the `LB_parameters_gpu` struct holds the lattice dimensions, but the `LB_parameters` (for cpu) does not, so sometimes it is fetched from the `lattice` global variable:
https://github.com/espressomd/espresso/blob/b24864f4a56ce05356b898e51cfbaa53612f417b/src/core/grid_based_algorithms/lb_interface.cpp#L501
and other times it is calculated without rounding:
https://github.com/espressomd/espresso/blob/b24864f4a56ce05356b898e51cfbaa53612f417b/src/core/grid_based_algorithms/lb_interface.cpp#L648-L652
even though the `lattice` object uses rounding:
https://github.com/espressomd/espresso/blob/b24864f4a56ce05356b898e51cfbaa53612f417b/src/core/grid_based_algorithms/lattice.cpp#L39
Is there a reason for not using the lattice object directly in the CPU code?

Commit:
Fixed issue in saving ASCII checkpoint file

The line change was previously done after 19 populations causing
overlap of values. It is now changed to be done after each value

Commit:
Test LB checkpointing at all lattice sites

The precision for GPU test is kept at 5 whereas for CPU it is 9.

jngrad committed:
Merge pull request #3 from jainka/extend_lb_checkpointing

Extend LB checkpointing tests

jngrad commented:
Container `check_with_odd_no_of_processors` says the CPU checkpoint file cannot be written to disk. Can't reproduce the issue locally inside the Docker container. Not sure why MPI should interact here, the only change in the save function for the CPU LB was:
https://github.com/espressomd/espresso/blob/98c60a40634882ff55bc1cd34f2b3dfcf92f5c42/src/core/grid_based_algorithms/lb_interface.cpp#L794-L799

jngrad committed:
Test file creation in LB checkpoint save test

KaiSzuttor commented:
@jngrad I don't think that there is a reason for not using ```lblattice``` directly. I just missed that while refactoring the CPU LB.

jngrad commented:
@KaiSzuttor thanks! I'll make the change. I've also reproduced the traceback from the failing container. The issue shows up when adding a CPU LB to the actor list and running the script with `mpiexec -n 3`:
```python
lbf = espressomd.lb.LBFluid(agrid=0.5, visc=1.3, dens=1.5, tau=0.01)
system.actors.add(lbf)
```
The failure is at this line, for n = 0:
https://github.com/espressomd/espresso/blob/b24864f4a56ce05356b898e51cfbaa53612f417b/src/core/grid_based_algorithms/lbboundaries.cpp#L278

Have you ever seen this  MPI interaction before?

fweik commented:
Have you checked that `lb_fields` is actually allocated?

KaiSzuttor committed:
Merge branch 'python' of https://github.com/espressomd/espresso into fix-2436

KaiSzuttor commented:
MnotWE:
``` python
import espressomd
import espressomd.lb

system = espressomd.System(box_l=[10.0, 12.0, 14.0])
system.cell_system.skin = 0.1
system.time_step = 0.01
lbf = espressomd.lb.LBFluid(agrid=0.5, visc=1.3, dens=1.5, tau=0.01)
system.actors.add(lbf)                                              
```

KaiSzuttor commented:
not yet clear to me, but the issue seems to be located in ```mpi_bcast_lb_params``` which is called when setting ```agrid```

KaiSzuttor commented:
ahh, ```agrid``` is incompatible with the local box

KaiSzuttor commented:
@fweik how to handle this? there is only a check in the lattice class that prints to ```runtimeErrorMsg```

KaiSzuttor committed:
testsuite: fix checkpoint test on 3 nodes.

fweik commented:
This would be the reason why most codes let the user choose the grid dimensions and calculate the grid constant and not the other way around. Not sure if we want to change this, but since we just made incompatible changes to the LB interface anyway now would be the time. What is the opinion of our UX officer @RudolfWeeber on this?

jngrad commented:
Ah ok, it was my understanding that espresso threw an error for `agrid` values incompatible with the number of nodes. For example using the MnotWE, with 5 nodes: `Exception: LB fluid activation: ERROR: box_l in direction 0 is too small`, with 4 nodes: runs ok, with 3 nodes: MPI traceback. Thanks! I would never have suspected the box size to be the issue.

jngrad commented:
Also, I get a similar MPI traceback when the box dimensions are not integer multiples of `agrid`; e.g. `agrid=0.50001`. This means floating point errors can cause a failure, too.

fweik commented:
@KaiSzuttor is also think there should be bounds checking with `ADDITIONAL_CHECKS` enabled.

fweik commented:
@jngrad yes, this is why the user should chose the integral value and not the fractional one.

KaiSzuttor commented:
if we'd the time to rewrite the lattice class and the interface...

jngrad committed:
Repair LB `agrid` error handler, and test it

jngrad committed:
Re-use known lattice size; test a right cuboid box

jngrad commented:
@KaiSzuttor the sanity check should now be doing its job.

RudolfWeeber commented:
Ready to merge?

KaiSzuttor committed:
Merge branch 'python' of https://github.com/espressomd/espresso into fix-2436

jngrad commented:
@RudolfWeeber there are still open points in https://github.com/espressomd/espresso/issues/2555#issuecomment-471038543

jngrad commented:
For reference: this PR also fixes #1848, fixes #2555, fixes #2564

KaiSzuttor commented on src/core/grid_based_algorithms/lattice.cpp:
```diff
@@ -45,17 +45,22 @@ int Lattice::init(double *agrid, double *offset, int halo_size, size_t dim) {
     }
   }
+  if (incompatible_agrid) {
+    return ES_ERROR;
```
I think that does not help, since the return value is not checked... we could also directly return in after the ```runtimeErrorMsg```

jngrad replied:
Thanks! Both the `.at(n)` solution and the return value check solution worked; we now get a clean `Exception: LB fluid activation` instead of the full MPI traceback.

jngrad committed:
Improve error handling

jngrad commented:
@RudolfWeeber it's ready to merge
